src/chat/MarkAsReadButton.js

```diff
@@ -55,6 +62,23 @@ export default function MarkAsReadButton(props: Props): Node {
     }
   }, [auth, narrow, streams]);
 
+  const markPmAsRead = useCallback(() => {
+    const userIds = userIdsOfPmNarrow(narrow);
+    let messageIds = undefined;
+    const key = pmUnreadsKeyFromPmKeyIds(userIds, ownUserId);
+    if (userIds.length > 1) {
+      messageIds = unreadHuddles.find(x => x.user_ids_string === key)?.unread_message_ids ?? [];
+    } else {
+      messageIds = unreadPms.find(x => x.sender_id.toString() === key)?.unread_message_ids ?? [];
+    }
```

Comment:
Yikes, this is kind of messy, isn't it. Particularly that gotcha commented on in the first commit.

I'd like to have this kind of code, dealing with the messy structure inside a data model, be located inside that model's code. So e.g. a `getUnreadIdsForPmNarrow`, located among the selectors at the top of `unreadModel.js`.

There's an existing spot where we do something similar, in `getUnreadCountForNarrow` in `unreadSelectors.js`:
```js
      pm: ids => {
        if (ids.length > 1) {
          const unreadsKey = pmUnreadsKeyFromPmKeyIds(ids, ownUserId);
          const unreadItem = unread.huddles.find(x => x.user_ids_string === unreadsKey);
          return unreadItem?.unread_message_ids.length ?? 0;
        } else {
          const senderId = ids[0];
          const unreadItem = unread.pms.find(x => x.sender_id === senderId);
          return unreadItem?.unread_message_ids.length ?? 0;
        }
      },
```

So a good structure would be to factor that logic out of there, and then that case of `getUnreadCountForNarrow` just looks like `ids => getUnreadIdsForPmNarrow(ids).length`.

Note that that also has a variation of how to deal with the string/number thing, that's perhaps a bit cleaner.

Reply:
> Note that that also has a variation of how to deal with the string/number thing, that's perhaps a bit cleaner.

Yeah, maybe. It's not a biggie, but that one doesn't use `pmUnreadsKeyFromPmKeyIds` at all in the 1:1 pm case, even though it's written to be able to handle it.